### DISCUSSION TITLE
Setting up automated tests for optional third party library

### DISCUSSION THREAD

**Discussion by rks171** [idaholab/moose#20190] [Q&A Tools] [answered]
I have created an optional dependency on a third party library (TPL) in the MOOSE app I am working with.  The way in which I did this is:

1. Compile the TPL separate from my MOOSE app but using the same env as was used for MOOSE compilation
2. Modify the MOOSE app Makefile to add the TPL lib/include dirs to the ADDITIONAL_INCLUDES and ADDITIONAL_LIBS variables and also define the TPL name in ADDITIONAL_CPPFLAGS.
3. Create a class in my MOOSE app for interfacing with this TPL.  Ifdef statements are used to only include the TPL calls if the TPL variable was defined in ADDITIONAL_CPPFLAGS.

This allows me to use the TPL successfully as long as it was included in the build.  Now I would like to add some automated tests using the MOOSE testrunner to protect this feature.  However, I need for the test to only be run if the TPL was included during the build and ignored completely otherwise.  Is there a way to do this?

**Comment by GiudGiud** [comment]
Hello

You need to set up a test group for automated testing. All the tests that will use the TPL will have to be in that group, and when you want to run them you will have to type:
```
./run_tests -g group_name_with_tpl
```
Add in the tests file for each concerned test:
`group = 'group_name_with_tpl' `

Guillaume

**Reply by rks171** [reply]
I just tested this and this does allow me to specifically run these tests; however, when I simply type:

```./run_tests```

the test runner is still picking them up.  This is going to cause these tests to fail if the TPL was not made available during the build.  Is there a way to ensure that group does not get run when the TPL was not made available?

**Reply by GiudGiud** [reply]
Well you'll have to be running `./run_tests --not_group <group_name>`

Or you can hack and mark the tests as `heavy`. Then they will only be ran with `--heavy`

Or you can place them in a different folder `tests_tpl` and only run them with this option:
```
 -C dir, --test-root dir
                        Tell the TestHarness to search for test spec files at
                        this location.
```

**Reply by rks171** [reply]
Thanks, on that last suggestion, isn't the default behavior of the run_tests script to walk the entire directory structure to find any directories that include a `tests` file?   So even moving these to their own directory will still result in them being picked up.  Unless I'm putting this new subdirectory in the wrong place?

**Reply by GiudGiud** [reply]
not sure. 
It shouldnt pick them up in your_app/
it may pick them up in your_app/test/

**Reply by rks171** [reply]
I tested moving these tests out of my `yourapp/test` dir and into the root directory, `yourapp`, and the runner still finds the tests, so it must just be searching everything from the root down.

**Reply by GiudGiud** [reply]
Umm. Well I understand you are trying to keep the standard behavior to be `./run_tests` and not `./run_tests` + some option but I m running out of ideas. 
@loganharbour anything I missed to be able to exclude a group of tests by default?

**Reply by loganharbour** [reply]
I wouldn't suggest groups here. We should just add the dependency as a test spec option that we can check for. That way, @rks171 can do something like `tpl=true` in his spec.

**Reply by rks171** [reply]
@loganharbour, are you saying this would need to be added as a new feature to the test runner?  And would it work by the caller passing in some flag to the `run_tests` script?

**Reply by loganharbour** [reply]
Yes, that's correct

**Comment by rks171** [comment, marked as answer]
I want to point out that there is another solution for this using submodules.  If the TPL is pulled into the MOOSE app as a git submodule, then any tests that are defined that require that TPL can have the following input added to the test definition:

```
required_submodule = 'submodule name'
```

If the submodule is present, the tests will be run, and if not, they will be skipped.
